gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspSubplugin.kt

```diff
@@ -73,6 +73,29 @@ import java.util.concurrent.Callable
 import javax.inject.Inject
 import kotlin.reflect.KProperty1
 
+@Suppress("INVISIBLE_REFERENCE", "INVISIBLE_MEMBER", "EXPOSED_PARAMETER_TYPE")
+internal class Configurator : AbstractKotlinCompileConfig<AbstractKotlinCompile<*>> {
+    constructor(compilation: KotlinCompilationData<*>, kotlinCompile: AbstractKotlinCompile<*>) : super(compilation) {
+        configureTask { task ->
+            if (task is KspTaskJvm) {
+                // Assign moduleName different from kotlin compilation to
+                // work around https://github.com/google/ksp/issues/647
+                // This will not be necessary once https://youtrack.jetbrains.com/issue/KT-45777 lands
+                task.moduleName.value(kotlinCompile.moduleName.map { "$it-ksp" })
+            }
+            if (task is KspTaskJS) {
+                val libraryCacheService = project.rootProject.gradle.sharedServices.registerIfAbsent(
+                    "${Kotlin2JsCompile.LibraryFilterCachingService::class.java.canonicalName}" +
+                        "_${Kotlin2JsCompile.LibraryFilterCachingService::class.java.classLoader.hashCode()}",
```

Comment:
Nit: `Kotlin2JsCompile.LibraryFilterCachingService::class.java`, which seems a bit long, is repeated. You can add a local variable for that `Class`.